### PR TITLE
Dstore optimizations

### PR DESCRIPTION
**Pull request by karasevb**
1. Added support for multiple dstore format module.
2. Optimize key search:
2.1. Use flags to identify Extension and Invalid 
2.2. Use hash function as a faspath for non-matching strings strings
3. Remove extra extention slots. 
4. Remove extra mallocs.

### CODE REVIEW

**Comment by karasevb** [comment]
image

**Comment by karasevb** [comment]
@artpol84 please review

File: src/client/pmix_client_get.c [hunk 1]
```diff
@@ -83,14 +83,6 @@ PMIX_EXPORT pmix_status_t PMIx_Get(const pmix_proc_t *proc, const char key[],
...
     pmix_status_t rc;
 
-    PMIX_ACQUIRE_THREAD(&pmix_global_lock);
-
-    if (pmix_globals.init_cntr <= 0) {
-        PMIX_RELEASE_THREAD(&pmix_global_lock);
-        return PMIX_ERR_INIT;
-    }
-    PMIX_RELEASE_THREAD(&pmix_global_lock);
-
```
Comment (rhc54) [inline src/client/pmix_client_get.c]: I take it you are then relying on checking for initialization in the non-blocking call? If so, we probably should review the code that executes prior to calling the nb function to ensure that nothing there relies on PMIx being initialized

Comment (rhc54) [reply src/client/pmix_client_get.c]: Actually, you don't have to look far - line 95 (in your changed version) calls PMIX_NEW which requires that the object system be initialized. I'm afraid the lock has to remain.

Comment (karasevb) [reply src/client/pmix_client_get.c]: This change has been rejected.

**Comment by rhc54** [comment]
@karasevb It occurs to me that one could replace that PMIX_ACQUIRE_THREAD code with an atomic lock, which might (should?) be faster. We would have to do it everywhere we check for initialization, but I _believe_ we don't have to worry about thread stall here as we aren't engaging in any operations (just checking the init flag).

File: src/mca/gds/ds12/gds_ds12_base.c [hunk 2]
```diff
@@ -23,15 +23,22 @@
...
 static pmix_common_dstore_ctx_t *ds12_ctx;
 
 static pmix_status_t ds12_init(pmix_info_t info[], size_t ninfo)
 {
     pmix_status_t rc = PMIX_SUCCESS;
+    pmix_common_dstore_file_cbs_t *dstore_file_cbs = NULL;
 
-    ds12_ctx = pmix_common_dstor_init(&pmix_ds12_lock_module, "ds12", info, ninfo);
+    if (!PMIX_PROC_IS_SERVER(pmix_globals.mypeer)) {
+        dstore_file_cbs = &pmix_ds20_file_module;
```
Comment (artpol84) [inline src/mca/gds/ds12/gds_ds12_base.c]: Why no ds12 here?

File: src/mca/common/dstore/dstore_file.h [hunk 3]
```diff
@@ -0,0 +1,161 @@
+#ifndef DSTORE_FORMAT_H
+#define DSTORE_FORMAT_H
+
+typedef size_t (*pmix_common_dstore_kv_size_fn)(uint8_t *addr);
+typedef char* (*pmix_common_dstore_key_name_ptr_fn)(uint8_t *addr);
+typedef size_t (*pmix_common_dstore_key_name_len_fn)(char *key);
+typedef uint8_t* (*pmix_common_dstore_data_ptr_fn)(uint8_t *addr);
+typedef size_t (*pmix_common_dstore_data_size_fn)(uint8_t *addr, uint8_t* data_ptr);
+typedef size_t (*pmix_common_dstore_key_size_fn)(char *key, size_t data_size);
+typedef size_t (*pmix_common_dstore_slot_size_fn)(void);
```
Comment (artpol84) [inline src/mca/common/dstore/dstore_file.h]: Is this for the extension slot? If so this should be reflected in the name of a function

File: src/mca/gds/ds12/gds_ds12_base.c (hunk 2 above)
Comment (karasevb) [reply src/mca/gds/ds12/gds_ds12_base.c]: The ds12 module has two dstore formats for 12 and 20. By default, it uses v20 format for dstor, but if a client will v12 then it is changed to v12.

File: src/mca/common/dstore/dstore_file.h (hunk 3 above)
Comment (karasevb) [reply src/mca/common/dstore/dstore_file.h]: Updated.

**Comment by artpol84** [comment]
@karasevb what is the status of this?

**Comment by rhc54** [comment]
Is this PR no longer required?

**Comment by karasevb** [comment]
We have discussed with @artpol, this PR should go. I will resolve merge conflicts.

**Comment by artpol84** [comment]
@rhc54 we will need this to be released along with the new dstor component as these changes are changing the format of the dstor.
Has "ds21" component been already released as part of 2.1 or 3.0?

**Comment by artpol84** [comment]
I see that 2.2 and 3.0 are @ the RC stages.
Would want to go add it there.

**Comment by karasevb** [comment]
Please do not merge until I complete the double check with MTT.

**Comment by karasevb** [comment]
Ready to go.

**Comment by jjhursey** [comment]
@karasevb Can you add a high level description of the optimization contained in this PR? Can you add it to the PR description so it makes it into the git logs?

The performance graph shows improvement, but I don't know what the design change was.

**Comment by karasevb** [comment]
@jjhursey Done. The description of PR has been updated.

**Comment by jjhursey** [comment]
Have you all checked the cross-version compatibility for this change?

It's fine to merge into master (I'll let @artpol84 review/merge), but before we merge to the release branches we need an assessment of cross-version compatibility.

**Comment by artpol84** [comment]
@jjhursey thanks for not merging it.
I wanted to check additionally commit that avoid extra extension slots and I think though it is correct it is more a hack to unroll undesired behaviour rather than a root-cause fix.
I'm going to prepare a fix now and will open a new PR with the set of commits from this PR + a proper extension slot fix.

**Comment by rhc54** [comment]
> @jjhursey do we have an automatic way of checking cross-version compatibility?
> As we only change the format of the ds21 we are not expecting to break it, but it would be good to double-check this.

We have the following script: https://github.com/pmix/pmix-tests/blob/master/crossversion/xversion.py. Would love to see it in MTT tests, though we might want to update it to work with posted tarballs or some other means of detecting that a branch changed 😄

**Comment by karasevb** [comment]
Closed due #908

**Pull request closed without merging**